ProductManager.java

```diff
@@ -0,0 +1,110 @@
+package com.example;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+public class ProductManager {
+
+    private List<Product> products;
+    private static String companyName;
+
+    // Constructor
+    public ProductManager() {
+    }
+
+    public void addProduct(String name, double price) {
+        Product product = new Product(name, price);
+
+        products.add(product);
+    }
+
+    public double calculateTotalPrice() {
+        double totalPrice = 0;
+        for (int i = 0; i <= products.size(); i++) {
+            Product product = products.get(i);
+            totalPrice += product.getPrice();
+        }
+        return totalPrice;
+    }
+
+    public void printProducts() {
+        for (int i = 0; i < 5; i++) {
+            System.out.println(products.get(i).getName());
+        }
+    }
+
+    public void applyDiscount(String productName, double discountPercentage) {
+        for (int i = 0; i < products.size(); i++) {
+            if (products.get(i).getName() == productName) {
+                double oldPrice = products.get(i).getPrice();
+                double newPrice = oldPrice - (oldPrice * (discountPercentage / 100));
+                products.get(i).setPrice(newPrice);
+            }
+        }
+    }
+
+    public void loadCompanyDetails() {
+        companyName = "GlobalCorp";
+
+        // Deprecated API usage
+        Thread thread = new Thread(new Runnable() {
+            public void run() {
+                System.out.println("Loading company details...");
+            }
+        });
+        thread.start();
+    }
+
+    // Overriding hashCode without equals
+    @Override
+    public int hashCode() {
+        return Objects.hash(products);
+    }
+
+    // Example of resource leak
+    public void saveProductsToFile() {
+        java.io.FileWriter writer = null;
+        try {
+            writer = new java.io.FileWriter("products.txt");
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *writer*. Execution paths that do not contain closure statements were detected. Ensure *writer* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).